### PR TITLE
Typescript rewrite

### PR DESCRIPTION
**Pull request by GrandSchtroumpf**
Add : 
- `tsconfig.json`.
- `package.json` : `@types/node` + `@types/bn.js`.
- `test/index.js` : change RLP directory to `dist/index.js`.
- src : typescript files.

TODO:
- [x] tests in TS
- [x] linting
- [x] prettier
- [x] make sure that it works with: https://github.com/ethereumjs/rlp/issues/35
- [x] remove original JS sources
- [x] Remove dependency on `yarn`
- [x] Update ``README``
- [x] Make sure that package importing works (entrypoint etc)
- [x] Extract common config
- [x] figure out `BigNumber` imports **SKIPPED** (note by @krzkaczor) so current typings are pretty shitty and they forces us to use it in this way. We could fork/create our package. I already done so in [evm-ts](https://github.com/ethereum-ts/evm-ts/blob/master/lib/%40types/bn.js/index.ts). This allows to use normal syntax for import. I would say it's low priority for now 🤷‍♂️ )

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+// Check is input is a number
+function isNumber(input: RLPInput): input is number {
+    return typeof input === 'number';
+}
+
+// Check if an input is a BigNumber
+function isBN(input: RLPInput): input is BN {
+    return !!input.hasOwnProperty('toArray');
```
Comment (axic) [inline src/index.ts]: bn.js has a function to do this: `BN.isBN(instance)`

**Comment by holgerd77** [comment]
Hi @GrandSchtroumpf, great initial work, thanks so much, also cool to see the tests already passing! 👍 We can leave this open for some time for some further tooling discussion and for improvements from other PRs integrated here, seen that @krzkaczor has already started with https://github.com/ethereumjs/rlp/pull/38.

**Comment by holgerd77** [comment]
Probably a linting update PR (we had a strong tendency for [Prettier](https://prettier.io) for this can also be done by @krzkaczor, so that tests are passing?

**Comment by holgerd77** [comment]
Some addition to using TypeScript ``strict``: I also had some discussion with Matthew @mattdean-digicatapult (who is currently doing great [StateManager](https://github.com/ethereumjs/ethereumjs-vm/pull/376) refactoring work on the VM) on this: Matthew was also very supportive of making type checking as strict as possible, but he also pointed out that this could make initial transition effort significantly larger since there can be the need of a lot more changes to the original JS code compared to a non-strict approach (depending very much on the structure of the initial code base). So it might make sense here to loosen this go-for-strict approach on a first round, just for the sake of not overloading on changes, and target this on a second update round.

But we can very much see how this is turning out on practical experiences and then decide how to proceed library-by-library.

**Comment by krzkaczor** [comment]
@holgerd77 yeah i think thats a valid point but this is very small lib so i guess it wasn't difficult to do it at one go.

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+  }
+
+  return decoded.data;
+}
+
+/**
+ * Get the length of the RLP input
+ * @param input
+ * @returns The length of the input or an empty Buffer if no input
+ */
```
Comment (holgerd77) [inline src/index.ts]: Is this intended to actually remove the types from the comments here? Does TypeScript do some auto-magic here?

**Comment by holgerd77** [comment]
@krzkaczor Whew, that was a quick merge! 😄 We can do this for non-master branch related merges. Just want to make sure and for others to note: this is not setting things in stone, just providing a more unified ground for further discussion (if needed).

File: src/index.ts (hunk 2 above)
Comment (GrandSchtroumpf) [reply src/index.ts]: yes. tsdoc knows the type from the code. It would throw a warning if I didn't remove it.

Comment (holgerd77) [reply src/index.ts]: Ah ok, thanks for clarification, just also had a look in the TSDoc playground. I find it a bit strange that types then seems to be not part of the generated documentation and not completely getting the concept behind it. But if this is the official documentation format, so be it.

**Comment by GrandSchtroumpf** [comment]
Tomorrow yes.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+      return Buffer.from(v.toArray());
+    } else {
+      throw new Error("invalid type");
+    }
+  }
+  return v as Buffer;
+}
+
+// Check is input is a number
+function isString(input: RLPInput): input is string {
```
[the remark this answers is not in the document]
Comment (GrandSchtroumpf) [reply src/index.ts]: I cannot find it for the browser though.

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      "outDir": "./dist",
+      "sourceMap": false,
+      "declaration": false,
+      "module": "commonjs",
+      "moduleResolution": "node",
+      "emitDecoratorMetadata": false,
+      "experimentalDecorators": true,
+      "allowSyntheticDefaultImports": true,
+      "strict": true,
+      "noImplicitAny": true,
```
[the remark this answers is not in the document]
Comment (GrandSchtroumpf) [reply tsconfig.json]: `"allowSyntheticDefaultImports": true,` is needed for BN to be imported the ES6 way.

File: package.json [hunk 5]
```diff
@@ -42,10 +42,11 @@
     "@types/mocha": "^5.2.5",
     "@types/node": "^10.12.2",
     "bn.js": "^4.11.1",
-    "nyc": "^13.1.0",
+    "conditional-type-checks": "^0.5.0",
```
Comment (holgerd77) [inline package.json]: This is a very low-star library (26 or so) and has a mid-download rate on npm often being < 100. For maintenance uncertainty as well as security reasons I wouldn't want to have this integrated as a standard dependency for the library.

Let's please stick to dependencies which are to some extend well-established in the ecosystem.

Comment (holgerd77) [reply package.json]: Ah, just seeing that you are solving an issue with this. Hmm, any other way to achieve this? So I would get a bit more open to this then, but concerns remain.

Can we generally switch to have your ``typescript`` branch as a base branch and can you do further work on another branch as a PR towards this base branch? Then it get's easier to discuss further changes and things conflict less with the other work done in parallel.

Comment (GrandSchtroumpf) [reply package.json]: Yes, krzkaczor told me to do so on gitter. Sorry about that.
I think we can to it without this library. This was the "cleaner" way.

**Comment by holgerd77** [comment]
@GrandSchtroumpf I think you should generally remove (by rebase) your last commit, otherwise it will get super-messy for @krzkaczor to rebase on top of this (@krzkaczor: please correct me if I am wrong).

Can you just copy-paste your #35 issue fixing changes to somewhere, delete the last commit and then resubmit (without the semicolon changes, this is already introduced in the PR form Chris) as a separate PR?

File: src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (krzkaczor) [reply src/index.ts]: buuut isin't it a webpack job to provide it. This lib targets node so we should use whats there (as we already do with buffers etc). 
Buuuut turns out that its deprecated: https://nodejs.org/api/util.html#util_util_isstring_object 

i am not sure what we want to do this then. Usually I use lodash for such things but I am not sure if brining whole library is a good idea (maybe treeshaking makes its footprint minimal? @holgerd77 any thoughts?)

File: package.json (hunk 5 above)
Comment (holgerd77) [reply package.json]: No problem! 😄

If you haven't done much rebasing before, be careful, this is hairy. I am still in the process of always doing a local copy of the whole repo before every rebase, cause I am still always somewhat afraid.

File: src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (holgerd77) [reply src/index.ts]: This is a one-line internal function used on one (just checked) occasion in the library, why are we not getting rid of this altogether?

[the remark this answers is not in the document]
Comment (holgerd77) [reply src/index.ts]: (same for ``isNumber``)

**Comment by holgerd77** [comment]
Once you are ready please also push-force (``push --force``) the branch here with the last commit removed, than @krzkaczor can do the last changes, rebase on top of this and we can get this merged. Probably better to do this sooner than later since changes introduced there are pretty broad in scope.

**Comment by holgerd77** [comment]
Since we are slowly getting to some common ground on configuration and setting? Do you guys think it is viable to have some of this configuration extracted to an external dependency - ``ethereumjs-config`` or something, I think one of you once mentioned that?

How would this work regarding the integration and what config parts would be applicable to that? Do you have any example where this is done in practice and/or some article link on this?

**Comment by GrandSchtroumpf** [comment]
I don't want you to think I'm obsessed with that ^^, but I think this is exactly why git-submodules are good. You can extract all the config into the main module, and link submodule to it. I'll try to find some docs or examples.

**Comment by krzkaczor** [comment]
@holgerd77 extracting linting config is straightforward (tsconfig). There are some ideas how to reuse prettier cfg: https://github.com/prettier/prettier/issues/3146 

I think it totally makes sense to extract it to the common repo. I could make a PR this weekend, could you just create repo for me (or maybe i have permissons? 😆 idk)

**Comment by holgerd77** [comment]
@GrandSchtroumpf I actually DO think that you are a bit obsessed with that! 😄 Sorry, I think I can to a larger extend imagine how this would look like and think that this would create a monolithic monster, I have very much made up my mind not to follow this path. We can nevertheless further think along the lines of consolidating libraries where it makes sense.

**Comment by holgerd77** [comment]
@krzkaczor Very cool, have set up the following repo:
https://github.com/ethereumjs/ethereumjs-config

If you are doing a PR on this that would be great, then we could further discuss along the design choices made within the PR!

**Comment by GrandSchtroumpf** [comment]
@holgerd77 yeaah... I know, I think I'm a bit obsessed too ^^.

**Comment by holgerd77** [comment]
For now added some coverage thresholds for failure (85% absolute, -2.5% relative), so that CI should pass on next trigger.

**Comment by holgerd77** [comment]
Merged. 😄

**Pull request merged**